Allow running Cilium with kube-router for BGP routing

tgraf opened the pull request:
Fixes: #4446
Related: #2463

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4547)
<!-- Reviewable:end -->

ianvernon commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -70,7 +70,7 @@ cilium-agent
       --socket-path string                Sets daemon's socket path to listen for connections (default "/var/run/cilium/cilium.sock")
```
Typo in commit: "Also improve the default DaemonSet file to us" --> "Also improve the default DaemonSet file to use"

ianvernon commented on pkg/node/manager.go:
```diff
@@ -208,6 +208,37 @@ func replaceNodeRoute(ip *net.IPNet) {
 	}
```
Nit in commit: "This only makes sense when tunneling is enabled; when tunelling is disabled.."

ianvernon commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+Using kube-router to run BGP
+****************************
+
+This guide explains how to configure Cilium and kube-router to co-operate to
```
Provide a link to the kube-router homepage when first mentioning kube-router.

ianvernon commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+****************************
+
+This guide explains how to configure Cilium and kube-router to co-operate to
+use kube-router for BGP peering and route propagation and Cilium for policy
```
Nit: "for BGP peering and route propagation, and Cilium..."

ianvernon commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+
+.. note::
+
+    If you are running kube-route with overlay enabled, adjust for the
```
kube-router

ianvernon reviewed (dismissed):
Minor nits / doc changes recommended. Would also appreciate addition of tests which test the disabled tunnel mode, and possibly kube-router as part of the tunnels tests which are present in `test/k8sT/Tunnels.go`.

aanm commented on bpf/init.sh:
```diff
@@ -320,15 +323,10 @@ if [ "$MODE" = "vxlan" -o "$MODE" = "geneve" ]; then
-		rm $FILE
-	fi
+	# Remove eventual existing encapsulation device from previous run
+	ip link del cilium_vxlan 2> /dev/null || true
```
This will brake connectivity for existing containers, how can the user be aware of this?

joestringer replied:
The user is explicitly changing the tunnel mode, can they *not* be aware of this?

tgraf replied:
I don't think this is breaking compatibility. Tunneling mode is being removed. Are you worried about restoring endpoints? I think connectivity will break anyway because of MTU.

aanm commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+.. _kops_guide:
```
Is this supposed to be `kops_guide`?

tgraf replied:
It's a leftover, removed

aanm commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+
+.. note::
+
+    This feature is regarded as ``tech-preview``. Please provide feedback and
```
shouldn't we use `alpha` or `beta`?

tgraf replied:
I don't have a strong opinion as long as it is marked as pre stable. I've changed it to beta.

aanm commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+
+.. code:: bash
+
+    curl -LO https://raw.githubusercontent.com/cloudnativelabs/kube-router/master/daemonset/generic-kuberouter-only-advertise-routes.yaml
```
Please use a tag instead of `master`

https://raw.githubusercontent.com/cloudnativelabs/kube-router/v0.2.0-beta.7/daemonset/generic-kuberouter-only-advertise-routes.yaml

aanm replied:
not addressed

aanm commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+
+The following arguments are **optional** and may be set according to your
+needs. Please see the `kube-router user guide
+<https://github.com/cloudnativelabs/kube-router/blob/master/docs/user-guide.md>`_
```
Same here

aanm commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+
+    If you are running kube-route with overlay enabled, adjust for the
+    encapsulation overhead by setting the environment variable ``MTU`` to the
+    value 1450.
```
Where? How? Please indicate this to the users

tgraf replied:
This needs to be reworked anyway to use a new option. Will add clearer documentation.

aanm commented on Documentation/install/guides/kube-router.rst:
```diff
@@ -0,0 +1,116 @@
+In the above example, the routes in the form of ``10.2.0.0/24 dev
+tun-172052116`` are installed by kube-router whereas the route ``10.2.1.0/24
+via 10.2.1.18 dev cilium_host`` is installed by Cilium to handle the PodCIDR of
+the local node.
```
Can you provide a little more info here? What does this means? It means Cilium was running with `--tunnel=disabled` in all other nodes except `10.2.1.0/24`?

joestringer reviewed (commented):
Were you hitting MTU-related issues? I think that the defaults should still work; By default Cilium configures the CNI config with 1500B MTU, then when the Cilium CNI plugin reads this config, it always subtracts 50B (which should be enough for IP-in-IP) before installing routes into the endpoint. Then the other part is routes that Cilium creates for traffic sourced from the local node, to guide it towards other nodes. However, these routes are disabled by the functionality in this PR.

So what's left?

joestringer reviewed (approved):
BPF bits LGTM

tgraf committed:
agent: Add support for--tunnel=disabled option

Allow to disable tunneling by specifying the mode "disabled" without requiring
to also specify a device to attach the BPF program to (--device). This mode
establishes compatibility with components that provide connectivity by
configuring the Linux routing layer to route endpoint IPs, e.g. kube-router.

Also makes the tunnel mode configurable via the environment variable
CILIUM_TUNNEL.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Remove existing encapsulation devices when tunneling is disabled

Replace the complicated logic that tries to recover the encapsulation device
name and remove all possible encapsulation device names when tunneling is
disabled. The device gets re-created when tunneling is re-enabled.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Do not install node routes when tunneling is disabled

The behavior regarding route installation on each node so far has been to
always install a route for each node in the cluster to point to the local
cilium_host interface to force traffic to any endpoint CIDR IP via the Cilium
forwarding logic. This only makes sense when tunneling is enabled. When
tunneling is disabled, skip installation of these routes and delegate the route
installation to whatever component is in charge. This allows compatibility with
components such as kube-router in BGP mode.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
mtu: Don't account for encapsulation overhead if tunneling is disabled

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Provide guide on how to install Cilium + kube-router

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kubernetes: Configure tunnel mode via ConfigMap

Improve the default DaemonSet file to use the environment variable to derive
tunnel configuration. The default behavior is unchanged. Existing DaemonSet
files that specify the "-t" option will continue to prefer whatever value is
provided via the option.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Reject combination of --tunnel=disabled and --single-cluster-route

The options are not compatible. Reject the combination if specified.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Adjust SNAT conditions for direct routing mode

When not in tunneling mode, be more restrictive when to SNAT.

Ingress:
* Do not SNAT if source is in cluster range

Egress:
* Do not SNAT if destination is in cluster range

If load-balancing occured by kube-proxy, leave it up to kube-proxy

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Add ipcache based fallback to derive ingress source identity

The world identity is the catch-all bucket for all sources that cannot be
associated with a more specific identity. This can occur because the identity
is completely unknown or because the identity information could not be carried
in the packet respectively has been cleared on the way.

Fall back to using the ipcache to derive the source identity for all traffic
associated with the world identity to recover the identity for all IPs known to
Cilium.

This also allows to completely get rid of the remaining ingress CIDR pieces in
the datapath and all user space control pieces.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Simplify code by using viper directly

Signed-off-by: Thomas Graf <thomas@cilium.io>
